README.md
# Face-Detection-and-Recognition-system-python
Face Detection and Recognition system python with Python 3.6 and OpenCV 4.2
1.Start this project from UI.py
2.Create dataset by clicking new registration button.
3.Train this by clicking Train Model button
4.click on recognize button to start recognizing.
5.If face recognize successfully you can hide-unhide document, open dataset, show history,.
6.You can also give authentication permission to other user by recognizing authenticate user face. 

Note:1. Please download libraries accordingly by opening python file : opencv,tkinter,os,random,playsound,numpy,PIL,webbrowser,datetime
     2. Run it using python 3.6


History.txt
2019-04-13 16:53:24.899412
2019-04-13 16:54:07.973476
2019-04-13 16:56:08.236639
2019-04-13 16:56:35.580771
2019-04-13 16:58:05.951491
2019-04-13 17:00:07.354206
2019-04-13 17:01:54.613398
2019-04-13 17:04:40.850035
2019-04-13 17:06:19.897785
2019-04-13 17:08:27.598083
2019-04-13 17:14:45.679042
2019-04-13 17:15:32.633582
2019-04-13 17:16:15.205642
2019-04-13 17:17:04.507065
2019-04-13 17:18:48.365748
2019-04-15 10:56:04.041545
2019-04-15 10:58:00.872293
2019-04-15 10:58:44.399233
2019-04-15 11:00:19.609991
2019-04-15 11:01:05.355264
2019-04-15 11:05:29.394523
2019-04-15 11:06:19.215823
2019-04-15 11:09:01.239253
2019-04-15 11:09:47.385572
2019-04-15 11:11:49.634693
2019-04-15 11:19:53.255804
2019-04-15 11:24:26.470716
2019-04-15 11:25:00.786017
2019-04-16 09:47:06.491150
2019-04-16 09:48:27.462233
2019-04-16 19:57:30.913087
2019-04-16 19:59:46.875679
2019-04-16 20:02:41.548126
2019-04-16 20:03:46.859493
2019-04-16 20:04:26.994081
2019-04-16 20:06:12.801814
2019-04-16 20:08:02.379391
2019-04-16 20:10:38.884274
2019-04-18 19:34:17.660697
2019-04-18 19:35:46.305213
2019-04-20 17:51:58.479565
2019-04-20 17:53:04.117064
2019-04-20 17:53:59.444873
2019-04-20 18:27:52.742492
2019-04-20 18:32:14.027912
2019-04-20 18:33:09.913470
2019-04-20 21:39:39.946172
2019-04-20 21:40:46.924736
2019-04-20 21:41:35.594769
2019-04-22 08:24:18.970273
2019-04-22 08:25:52.842721
2019-04-22 08:27:07.981419
2019-04-22 09:18:08.448582
2019-04-22 09:31:09.682454
2019-04-22 09:47:19.294801
2019-04-22 09:49:32.869169
2019-04-22 09:51:21.042661
2019-04-22 09:55:39.910204
2019-04-22 09:56:29.474121
2019-04-22 09:58:39.692721
2019-04-22 09:59:00.263993
2019-04-22 09:59:48.093343
2019-04-22 10:03:54.468891
2019-04-22 10:05:11.772056
2019-04-22 10:05:25.534425
2019-04-22 10:05:54.666632
2019-04-22 10:06:26.215918
2019-04-22 10:06:44.462331
2019-04-22 10:07:40.170383
2019-04-22 10:09:51.395750
2019-04-22 10:10:14.485546
2019-04-22 10:11:21.487052
2019-04-22 10:13:21.775793
2019-04-22 10:30:10.176530
2019-04-22 10:30:28.599815
2019-04-22 10:31:34.235444
2019-04-22 10:32:45.179126
2019-04-22 10:33:44.770953
2019-04-22 10:34:53.805031
2019-04-22 10:36:34.491074
2019-04-22 17:20:53.167957
2019-04-22 17:21:33.881515
2019-04-22 17:22:06.283591
2019-04-22 17:27:36.006522
2019-04-22 17:28:22.649830
2019-04-22 19:41:42.718903
2019-04-22 19:43:12.502400
2019-04-22 19:44:12.172720
2019-04-22 19:44:26.871834
2019-04-22 19:49:25.475406
2019-04-22 19:50:01.012506
2019-04-22 19:50:31.137092
2019-04-22 19:51:51.737023
2019-04-22 19:52:04.296636
2019-04-22 19:52:34.351489
2019-04-22 19:52:48.584976
2019-04-22 19:53:06.900090
2019-04-22 19:53:57.566074
2019-04-22 19:54:26.803626
2019-04-22 19:54:40.482354
2019-04-22 19:54:57.043170
2019-04-22 19:55:08.978055
2019-04-22 19:55:26.234136
2019-04-22 19:55:44.585192
2019-04-22 19:56:37.190954
2019-04-22 19:59:16.491169
2019-04-22 19:59:56.849787
2019-04-23 09:08:10.147914
2019-04-23 09:23:07.937307
2019-04-23 09:24:18.797998
2019-04-23 09:24:38.901883
2019-04-23 09:25:02.586918
2019-04-23 09:31:46.848723
2019-04-23 09:32:51.127381
2019-04-23 09:34:29.071057
2019-04-23 09:34:45.113253
2019-04-23 09:43:06.602896
2019-04-23 09:43:54.339917
2019-04-23 09:44:09.342647
2019-04-23 09:47:21.747112
2019-04-23 09:47:42.308900
2019-04-24 09:18:54.543290
2019-04-24 09:21:52.029765
2019-04-24 09:26:27.715810
2019-04-24 09:26:43.319135
2019-04-24 09:30:16.782794
2019-04-24 09:32:31.084629
2019-04-24 09:34:43.881712
2019-04-24 09:35:16.169259
2019-04-24 09:35:39.230651
2019-04-24 09:37:52.324040
2019-04-24 09:41:05.663031
2019-04-24 09:51:33.241748
2019-04-24 09:59:15.524509
2019-04-24 10:01:30.432346
2019-04-24 10:01:45.730113
2019-04-24 10:27:34.457653
2019-04-24 10:32:07.007108
2019-04-24 10:32:26.923615
2019-04-24 10:32:55.898679
2019-04-24 10:33:25.196275
2019-04-24 10:34:25.247008
2019-04-24 10:50:59.186496
2019-04-25 08:55:19.911616
2019-04-25 19:19:33.661885
2019-04-25 19:20:46.987260
2019-04-25 19:21:51.963894
2019-04-25 19:23:45.191655
2019-04-27 13:53:26.009170
2019-04-27 13:55:47.602349
2019-04-27 13:56:07.825552
2019-04-27 13:56:22.842085
2019-04-27 13:57:09.791584
2019-04-29 16:59:52.593259
2019-04-29 17:00:42.119326
2019-04-29 17:01:02.831876
2019-04-29 17:01:29.786894
2019-04-29 17:11:57.927979
2019-04-29 17:12:20.900108
2019-04-29 17:13:03.696071
2019-04-29 17:15:10.880093
2019-04-29 17:15:32.245806
2019-04-29 17:30:58.349688
2019-04-29 17:33:37.999255
2019-04-29 17:34:06.060211
2019-04-29 17:37:36.493198
2019-04-29 17:38:15.553682
2019-04-29 17:39:05.878765
2019-04-29 17:39:15.411381
2019-04-29 17:39:54.624324
2019-04-30 10:34:06.249698
2019-04-30 10:35:41.270234
2019-04-30 10:36:36.598199
2019-04-30 10:36:59.514075
2019-04-30 10:37:45.355691
2019-04-30 10:39:35.038146
2019-04-30 10:40:58.170634
2019-05-02 09:42:48.633918
2019-05-02 09:44:14.293049
2019-05-02 11:03:14.436180
2019-05-02 21:22:31.698070
2019-05-02 21:23:03.481059
2019-05-02 21:23:28.950736
2019-05-03 08:47:16.571121
2019-05-03 08:47:36.207715
2019-05-03 08:49:28.910459
2019-05-03 08:49:34.438311
2019-05-03 08:50:54.516658
2019-05-03 15:44:24.656364
2019-05-03 16:28:31.551253
2019-05-03 16:29:48.566145
2019-05-03 17:51:04.729415
2019-05-03 17:51:55.401992
2019-05-03 20:11:47.353467
2019-05-03 20:13:14.204717
2019-05-03 20:13:28.636053
2019-05-03 21:55:20.860484
2019-05-04 06:33:37.198592
2019-05-04 06:34:12.849202
2019-05-04 06:34:54.425813
2019-05-04 07:44:02.508764
2019-05-04 07:44:54.794919
2019-05-04 07:45:35.730604
2019-05-04 07:46:44.376252
2019-05-04 07:56:15.253381
2019-05-04 09:17:29.824670
2019-05-04 09:20:17.544961
2019-05-04 09:20:37.766131
2019-05-04 09:21:09.800239
2019-05-04 09:22:09.907891
2019-05-04 09:22:32.793767
2019-05-04 09:23:12.338005
2019-05-04 09:26:15.871718
2019-05-04 10:19:56.227673
2019-05-04 10:20:20.247806
2019-05-04 10:20:41.569357
2019-05-04 10:22:02.385930
2020-08-30 11:53:04.960422
2020-08-30 11:53:23.563262
2020-08-30 11:54:27.098543
2020-08-30 11:55:42.647636
2020-08-30 12:01:06.573718

UI.py
import numpy
import time
import cv2
from tkinter import *
import os
from playsound import playsound


os.system("splash_screen1.py")
root=Tk()
root.geometry('500x570')
frame = Frame(root, relief=RIDGE, borderwidth=2)
frame.pack(fill=BOTH,expand=1)
root.title('Welcome To Face Detection System!!!')
frame.config(background='light blue')
label = Label(frame, text="Face Detection System",bg='light blue',font=('Times 35 bold'))
label.pack(side=TOP)
filename = PhotoImage(file="demo.png")
background_label = Label(frame,image=filename)
background_label.pack(side=TOP)


def function2():
    os.system("permission_for_registration.py")


def function3():
    os.system("lock_unlock_face_recognition.py")


def function4():
    os.system("training_model.py")

def function5():
    os.system("Project_Documentation.docx")

def function6():
    root.destroy()


menu = Menu(root)
root.config(menu=menu)

but1=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=function2,text='new registration',font=('helvetica 15 bold'))
but1.place(x=5,y=100)

but2=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=function4,text='Training Model',font=('helvetica 15 bold'))
but2.place(x=5,y=180)

but3=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=function3,text='Recognize',font=('helvetica 15 bold'))
but3.place(x=5,y=250)

but4=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=function5,text='Documentation of Project',font=('helvetica 15 bold'))
but4.place(x=5,y=320)

but5=Button(frame,padx=5,pady=5,width=5,bg='white',fg='black',relief=GROOVE,text='EXIT',command=function6,font=('helvetica 15 bold'))
but5.place(x=220,y=490)
playsound("sounds/welcome.mp3")

root.mainloop()



training_model.py
import cv2
import os
import numpy as np    #numpy for matrix calculations
from PIL import Image
from playsound import playsound

def assure_path_exists(path):           #same as in face_dataset.py
    dir = os.path.dirname(path)
    if not os.path.exists(dir):
        os.makedirs(dir)

#Now we need to initialize the recognizer and the face detector
recognizer = cv2.face.LBPHFaceRecognizer_create()       # Create Local Binary Patterns Histograms for face recognization
detector = cv2.CascadeClassifier("haarcascade_frontalface_default.xml")


#this will get the path of each images in the folder.
def getImagesAndLabels(path):       #method to get the images and label data
    imagePaths = [os.path.join(path,f) for f in os.listdir(path)]      # Get all file path
    faceSamples=[]         # create empty face sample list
    ids = []               # create empty id list

    for imagePath in imagePaths:     # Loop for all the file path
        PIL_img = Image.open(imagePath).convert('L')    # Get the image and convert it to grayscale
        img_numpy = np.array(PIL_img,'uint8')       # PIL image to numpy array
        id = int(os.path.split(imagePath)[-1].split(".")[1])      # Get the image id
        faces = detector.detectMultiScale(img_numpy)  # Get the face from the training images
        for (x,y,w,h) in faces:   # Loop for each face, append to their respective ID
            faceSamples.append(img_numpy[y:y+h,x:x+w])        # Add the image to face samples
            ids.append(id)   # Add the ID to IDs

    return faceSamples,ids

faces,ids = getImagesAndLabels('dataset/')  # Get the faces and IDs
recognizer.train(faces, np.array(ids)) # Train the model using the faces and IDss
assure_path_exists('trainer/')    # Save the model into trainer.yml
recognizer.save('trainer/trainer.yml')
playsound("sounds/trained.mp3")



dataset_capture.py
import cv2
import os

def assure_path_exists(path):
    dir = os.path.dirname(path)
    if not os.path.exists(dir):
        os.makedirs(dir)

face_id = input('enter your id:-')
# Start capturing video
vid_cam = cv2.VideoCapture(0)

# Detect object in video stream using Haarcascade Frontal Face
face_detector = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')

# Initialize sample face image
count = 0

assure_path_exists("dataset/")

# Start looping
while (True):

    # Capture video frame
    _, image_frame = vid_cam.read()

    # Convert frame to grayscale
    gray = cv2.cvtColor(image_frame, cv2.COLOR_BGR2GRAY)

    # Detect frames of different sizes, list of faces rectangles
    faces = face_detector.detectMultiScale(gray, 1.3, 5)

    # Loops for each faces
    for (x, y, w, h) in faces:
       # assure_path_exists("dataset/User." + str(face_id) + '.' + str(count))
        # Crop the image frame into rectangle
        cv2.rectangle(image_frame, (x, y), (x + w, y + h), (255, 0, 0), 2)

        # Increment sample face image
        count += 1

        # Save the captured image into the datasets folder
        cv2.imwrite("dataset/User." + str(face_id) + '.' + str(count) + ".jpg", gray[y:y + h, x:x + w])

        # Display the video frame, with bounded rectangle on the person's face
        cv2.imshow('frame', image_frame)

    # To stop taking video, press 'q' for at least 100ms
    if cv2.waitKey(100) & 0xFF == ord('q'):
        break

    # If image taken reach 30, stop taking video
    elif count >= 35:
        print("Successfully Captured")
        break

# Stop video
vid_cam.release()

# Close all started windows
cv2.destroyAllWindows()

create_face_dataset.py
#from email.contentmanager import set_text_content
import cv2  # import OpenCV
import os
from tkinter import *
import random
from playsound import playsound

ran=random.randint(1,1000)
def check_path(path):  # function to confirm whether the given path exists or not
    dir = os.path.dirname(path)  # if it doesn't exist this function will create
    if not os.path.exists(dir):
        os.makedirs(dir)


cam = cv2.VideoCapture(0)  # Start video capturing
face_cascade = cv2.CascadeClassifier('haarcascade_frontalface_default.xml')  # Detect object in video stream using Haarcascade Frontal Face
count = 0  # Initialize sample face image
check_path("dataset/")
playsound("sounds/prompt.mp3")
user=input("Enter name=")
while (cam.isOpened()):
    _, image_frame = cam.read()  # Capture video frame _, is used to ignored first value because vid_cam.read() is returning 2 values
    gray = cv2.cvtColor(image_frame, cv2.COLOR_BGR2GRAY)  # Convert frame to grayscale
    faces = face_cascade.detectMultiScale(gray, 1.4, 5)  # Detect faces using Cascade Classifier(xml file)

    for (x, y, w, h) in faces:
        cv2.rectangle(image_frame, (x, y), (x + w, y + h), (255, 0, 0), 2)  # Crop the image frame into rectangle
        count += 1  # Increment face image
        cv2.imwrite("dataset/" + str(user) + '.' +str(ran)+'.'+ str(count) + ".jpg",gray[y:y + h, x:x + w])  # Save the captured image into the datasets folder
        cv2.imshow('Creating Dataset!!!',image_frame)  # Display the video frame, with rectangular box on the person's face

    if cv2.waitKey(10) & 0xFF ==(' '):  # To stop taking video, press 'Esc'
        break

    elif count > 200:  # If image taken reach 100, stop taking video
        break

cam.release()  # Stop video
cv2.destroyAllWindows()  # Close all windows
playsound("sounds/samples.mp3")


permission_for_registration.py
import cv2
import os
from tkinter import *
from playsound import playsound
import os.path

root = Tk()
root.configure(background="blue")


def assure_path_exists(path):
    dir = os.path.dirname(path)
    if not os.path.exists(dir):
        os.makedirs(dir)


def check_path(path):  # function to confirm whether the given path exists or not
    dir = os.path.dirname(path)  # if it doesn't exist this function will create
    if not os.path.exists(dir):
        os.makedirs(dir)


recognizer = cv2.face.LBPHFaceRecognizer_create()
assure_path_exists("trainer/")
recognizer.read('trainer/trainer.yml')  #load training model
cascadePath = "haarcascade_frontalface_default.xml"  #cascade path
faceCascade = cv2.CascadeClassifier(cascadePath);  #load cascade
font = cv2.FONT_HERSHEY_COMPLEX_SMALL  # Set the font style

count = 0
count1 = 0
face_id = 1
check_path("dataset/")
path = "dataset/"
num_files = len([f for f in os.listdir(path)if os.path.isfile(os.path.join(path, f))])


if(num_files == 0):
    os.system("create_face_dataset.py")

else:
    playsound("sounds/permission2.mp3")
    cam = cv2.VideoCapture(0)
    while (cam.isOpened()):
        ret, im = cam.read()
        gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)
        faces = faceCascade.detectMultiScale(gray, 1.3, 5)
        for (x, y, w, h) in faces:
            Id, confidence = recognizer.predict(gray[y:y + h, x:x + w])  # Recognize the face belongs to which ID
            if (confidence <= 40):  # confidence usually comes greater than 80 for strangers
                cv2.rectangle(im, (x - 5, y - 5), (x + w + 20, y + h + 20), (0, 255, 0), 2)
                cv2.putText(im, str(round(confidence)), (x + 198, y + h), font, 1.00, (0, 255, 0), 1)
                print("allowed "+str(confidence))
                count = count + 1
                if (count == 15):
                    root.withdraw()
                    #tkinter.messagebox.showinfo('answer', 'You can now register')
                    playsound("sounds/permission granted.mp3")
                    cam.release()
                    os.system("create_face_dataset.py")

            else:  # confidence usually comes less than 80 for correct user(s)
                cv2.rectangle(im, (x - 5, y - 5), (x + w + 20, y + h + 20), (0, 0, 255), 2)
                cv2.putText(im, str(round(confidence)), (x + 198, y + h), font, 1.00, (0, 0, 255), 1)
                count1 = count1 + 1
                print("not allowed "+str(confidence))
                if (count1 == 200):
                    playsound("sounds/recognize unsuccessfully.mp3")
                    cam.release()

        cv2.imshow('Registered Face...', im)
        if cv2.waitKey(10) & 0xFF == ord(' '):  # If '*' is pressed, terminate the  program
            break
cv2.destroyAllWindows()


lock_unlock_face_recognition.py
from subprocess import call
import cv2
import sys
import numpy as np
import os
import datetime
from tkinter import *
import webbrowser
from playsound import playsound


#os.system("training_model.py")
now = datetime.datetime.now()
def assure_path_exists(path):
    dir = os.path.dirname(path)
    if not os.path.exists(dir):
        os.makedirs(dir)
        

recognizer = cv2.face.LBPHFaceRecognizer_create()
assure_path_exists("trainer/")
recognizer.read('trainer/trainer.yml')  #load training model
cascadePath = "haarcascade_frontalface_default.xml"  #cascade path
faceCascade = cv2.CascadeClassifier(cascadePath);  #load cascade
font = cv2.FONT_HERSHEY_COMPLEX_SMALL  # Set the font style


folderPath = 'D:/sample.docx'
def History():
    os.system("History.txt")


def hide():
        call(["attrib", "+H", folderPath])
        #tkinter.messagebox.showinfo('answer', 'HIDE SUCCESSFULLY....')
        playsound("sounds/hide.mp3")


def unhide():
        call(["attrib", "-H", folderPath])
        webbrowser.open('file:///' + folderPath)
        #tkinter.messagebox.showinfo('answer', 'UNHIDE SUCCESSFULLY....')
        playsound("sounds/unhide.mp3")

def dataset():
    Path='D:/BCA/bca-6/Face-Detection-and-Recognition-system-python/dataset'
    webbrowser.open('file:///' +Path)

def function6():
    root.destroy()

f = open("History.txt", "a")
f.write(str(now)+"\n")
f.close()

root=Tk()
root.geometry('500x570')
frame = Frame(root, relief=RIDGE, borderwidth=2)
frame.pack(fill=BOTH,expand=1)
root.title('Face Detection System')
frame.config(background='light blue')
label = Label(frame, text="Face Detection System",bg='light blue',font=('Times 35 bold'))
label.pack(side=TOP)
filename = PhotoImage(file="demo.png")
background_label = Label(frame,image=filename)
background_label.pack(side=TOP)

# stting title for the window
root.title("Face Comparision Found")

# creating second button
menu = Menu(root)
root.config(menu=menu)

but1=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=unhide,text='UNHIDE FILE',font=('helvetica 15 bold'))
but1.place(x=5,y=104)

but2=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=hide,text='HIDE FILE',font=('helvetica 15 bold'))
but2.place(x=5,y=190)

but3=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,command=dataset,text='Open Dataset Folder',font=('helvetica 15 bold'))
but3.place(x=5,y=270)

but6=Button(frame,padx=5,pady=5,width=39,bg='white',fg='black',relief=GROOVE,text='History',command=History,font=('helvetica 15 bold'))
but6.place(x=5,y=350)

but5=Button(frame,padx=5,pady=5,width=5,bg='white',fg='black',relief=GROOVE,text='EXIT',command=function6,font=('helvetica 15 bold'))
but5.place(x=220,y=490)

count=0
count1=0
cam = cv2.VideoCapture(0)
while (cam.isOpened()):

    ret, im =cam.read()
    if ret:
        gray = cv2.cvtColor(im, cv2.COLOR_BGR2GRAY)
        faces = faceCascade.detectMultiScale(gray, 1.3, 5)

        for (x, y, w, h) in faces:
            # cv2.rectangle(im, (x-20,y-20), (x+w+20,y+h+20), (0,255,0), 1)
            Id, confidence = recognizer.predict(gray[y:y + h, x:x + w])  # Recognize the face belongs to which ID

            if (confidence <= 40):  # confidence usually comes greater than 80 for strangers
                cv2.rectangle(im, (x - 5, y - 5), (x + w + 20, y + h + 20), (0, 255, 0), 2)
                cv2.putText(im, str(round(confidence)), (x + 198, y + h), font, 1.00, (0, 255, 0), 1)
                # Id, confidence = recognizer.predict(gray[y:y + h, x:x + w])  # Recognize the face belongs to which ID
                print("allowed   " + str(confidence))
                count = count + 1
                if (count == 15):
                    cam.release()
                    playsound("sounds/recognize successfully.mp3")
                    root.mainloop()

            else:  # confidence usually comes greater than 50 for correct user(s)
                cv2.rectangle(im, (x - 5, y - 5), (x + w + 20, y + h + 20), (0, 0, 255), 2)
                cv2.putText(im, str(round(confidence)), (x + 198, y + h), font, 1.00, (0, 0, 255), 1)
                # Id, confidence = recognizer.predict(gray[y:y + h, x:x + w])  # Recognize the face belongs to which ID
                count1 = count1 + 1
                print("not allowed  " + str(confidence))
                if (count1 == 250):
                    playsound("sounds/recognize unsuccessfully.mp3")
                    cam.release()

    cv2.imshow('Webcam',im) 

    if cv2.waitKey(10) & 0xFF == ord(' '):      # If space key is pressed, terminate the  program
        break
cam.release()
cv2.destroyAllWindows()
